builder/src/core/prefetch.rs

```diff
@@ -131,50 +146,63 @@ impl Prefetch {
             policy,
             disabled: false,
             patterns,
-            files: BTreeMap::new(),
+            files_prefetch: Vec::with_capacity(10000),
+            files_non_prefetch: Vec::with_capacity(10000),
         })
     }
 
-    /// Insert node into the prefetch list if it matches prefetch rules.
-    pub fn insert_if_need(&mut self, obj: &TreeNode, node: &Node) {
+    /// Insert node into the prefetch Vector if it matches prefetch rules,
+    /// while recording the index of matched prefetch pattern,
+    /// or insert it into non-prefetch Vector.
+    pub fn insert(&mut self, obj: &TreeNode, node: &Node) {
         // Newly created root inode of this rafs has zero size
         if self.policy == PrefetchPolicy::None
             || self.disabled
             || (node.inode.is_reg() && node.inode.size() == 0)
         {
+            self.files_non_prefetch.push(obj.clone());
             return;
         }
 
-        let path = node.target();
-        for (f, v) in self.patterns.iter_mut() {
-            // As path is canonicalized, it should be reliable.
-            if path == f {
-                if self.policy == PrefetchPolicy::Fs {
+        let mut path = node.target().clone();
+        let mut exact_match = true;
+        loop {
+            if let Some((idx, _, v)) = self.patterns.get_full_mut(&path) {
+                if exact_match {
                     *v = Some(obj.clone());
                 }
                 if node.is_reg() {
-                    self.files.insert(path.clone(), obj.clone());
+                    self.files_prefetch.push((obj.clone(), idx));
+                } else {
+                    self.files_non_prefetch.push(obj.clone());
                 }
-            } else if path.starts_with(f) && node.is_reg() {
-                self.files.insert(path.clone(), obj.clone());
+                return;
             }
+            // If no exact match, try to match parent dir until root.
+            if !path.pop() {
+                self.files_non_prefetch.push(obj.clone());
+                return;
+            }
+            exact_match = false;
         }
     }
 
-    /// Check whether the node is in the prefetch list.
-    pub fn contains(&self, node: &Node) -> bool {
-        self.files.contains_key(node.target())
-    }
+    /// Get node Vector of files in the prefetch list and non-prefetch list.
+    /// The order of prefetch files is the same as the order of prefetch patterns.
+    /// The order of non-prefetch files is the same as the order of BFS traversal of file tree.
+    pub fn get_file_nodes(&self) -> (Vec<TreeNode>, Vec<TreeNode>) {
+        let mut p_files = self.files_prefetch.clone();
+        p_files.sort_by_key(|k| k.1);
```

Comment:
[sort_unstable_by_key](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable_by_key) may help here:)

Reply:
I suggest to stable sort here, to keep the BFS iteration order for the files that match the same prefetch pattern index. E.g., for `/a` and `/b` that matching the prefetch pattern `/`, stable sort makes sure that `/a` dumps before `/b`.